tests/OAuthTest.php

```diff
@@ -0,0 +1,229 @@
+<?php
+
+namespace Tests;
+
+use App\Models\User;
+use App\Models\UserOAuthToken;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Config;
+use Laravel\Socialite\Facades\Socialite;
+use Mockery\LegacyMockInterface;
+use Mockery\MockInterface;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+class OAuthTest extends TestCase
+{
+    /** @var array|string[] The drivers we want to test */
+    protected array $drivers = ['discord'];
+
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        foreach ($this->drivers as $driver) {
+            Config::set('services.'.$driver.'.enabled', true);
+        }
+    }
+
+    /**
+     * Simulate what would be returned by the OAuth provider
+     *
+     * @return LegacyMockInterface|MockInterface
+     */
+    protected function getMockedProvider(): LegacyMockInterface|MockInterface
+    {
+        $abstractUser = \Mockery::mock('Laravel\Socialite\Two\User')
+            ->allows([
+                'getId'     => 123456789,
+                'getName'   => 'OAuth user',
+                'getEmail'  => 'oauth.user@phpvms.net',
+                'getAvatar' => 'https://en.gravatar.com/userimage/12856995/aa6c0527a723abfd5fb9e246f0ff8af4.png',
+            ]);
+
+        $abstractUser->token = 'token';
+        $abstractUser->refreshToken = 'refresh_token';
+
+        return \Mockery::mock('Laravel\Socialite\Contracts\Provider')
+            ->allows([
+                'user' => $abstractUser,
+            ]);
+    }
+
+    /**
+     * Try to link a logged-in user to an OAuth account from profile
+     *
+     * @return void
+     */
+    public function testLinkAccountFromProfile(): void
+    {
+        $user = User::factory()->create([
+            'name'  => 'OAuth user',
+            'email' => 'oauth.user@phpvms.net',
+        ]);
```

Comment:
You should also have a case where the email in the user doens't match the email or id for the user with the provider

Reply:
You mean the case where no user is found ?
It's there https://github.com/nabeelio/phpvms/pull/1735/files#diff-adfca0942be5b3a13d572d23bcaf199a28968818dbdadf9d8c3f375a7ebe8a65R157-R165

Reply:
This makes me think... Logically I mean;

Scenario;

* I have a discord account with myname@coldmail.com 
* I have a membership to a phpvms v7 powered VA with mysecondmail@zoomail.com 
* I have my email verified at phpvms v7

What will be response of the code, what should we do ?

* Should we update user's email ?
* Should we record user's external service provided data somewhere in v7 ? 
  _(like in oAuth tokes table, or somewhere else)_
* Should we simply disregard the difference, link accounts and let them in ?

Reply:
Here's how it's currently handled (the approach seems good to me):

It's not possible to link your account via the login page because there's no match at the email level, but it can be linked on the profile page (once logged in). 

I think this is the best way to handle it because a person might have used an old email when creating their Discord account, an email that is no longer active today, for example, and now they have an active email for the v7 where they want to receive notifications.

I don't see much benefit in forcing the user to change their email in phpVMS. As for storing the other email, I'm not sure if that's necessary either. We'll probably never use that other email, so it's best to avoid storing such personal data if we can.

Reply:
No forcing, no auto update... I am ok with that. 

So in this case the last choice is active `Disregard the difference, just link accounts` 😉